src/reconcile/rec_visibility.c

```diff
@@ -363,17 +391,18 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins,
 	 * order), so we track the maximum transaction ID and the newest update
 	 * with a timestamp (if any).
 	 */
-	timestamp = first_ts_upd == NULL ? 0 : first_ts_upd->durable_ts;
-	all_visible = upd == first_txn_upd &&
+	all_stable = upd == first_txn_upd &&
 	    !list_prepared && !list_uncommitted &&
-	    (F_ISSET(r, WT_REC_VISIBLE_ALL) ?
-	    __wt_txn_visible_all(session, max_txn, timestamp) :
-	    __wt_txn_visible(session, max_txn, timestamp));
+	    __wt_txn_visible_all(session, max_txn, max_ts);
 
-	if (all_visible)
+	if (all_stable)
 		goto check_original_value;
 
-	r->leave_dirty = true;
+	if (upd != first_txn_upd ||
+	    list_prepared || list_uncommitted ||
+		F_ISSET(r, WT_REC_VISIBLE_ALL) ||
+	    !__wt_txn_visible(session, max_txn, max_ts))
+		r->leave_dirty = true;
```

Comment:
This is the bug fix for the original failure that sparked all of this.  If a checkpoint is doing eviction, it runs with a timestamp, so `WT_REC_VISIBLE_ALL` is not set, but it's not okay to evict a page and discard its history when everything is visible to checkpoint -- there may be an older reader active (or equivalently we may be required to maintain history back to the application-specified `oldest_timestamp`).

With this change, if everything on a page is visible to checkpoint eviction but something is not `visible_all`, the page will still be left clean, but the updates will be saved, which will trigger saving the update to lookaside.